### PR TITLE
first steps towards 4.0

### PR DESCRIPTION
**Pull request by chives**
TODO in subsequent PRs:
- [x] replace `fsi/doctrine-extensions` with `fsi/files`
- [x] replace `fsi/datasource` and `fsil/datagrid` with `fsi/data`
- [ ] add strict types wherever possible
- [ ] make compatible with php `8.1`
- [ ] integration tests for positionable actions
- [ ] integration tests for tree reordering actions
- [ ] make admin events to be compatible with single argument `dispatch` method
- [ ] prepare `UPGRADE-4.0.md` document

### CODE REVIEW

File: Admin/Manager.php [hunk 1]
```diff
@@ -11,18 +11,17 @@
...
     /**
      * @var Element[]
      */
     protected $elements;
 
-    public function __construct()
+    public function __construct(iterable $visitors)
     {
         $this->elements = [];
+        $this->accept($visitors);
```
Comment (szymach) [inline Admin/Manager.php:24]: Perhaps we could postpone this until the first relevant method call (has/getElement(s))?

File: DependencyInjection/Compiler/ManagerVisitorPass.php [hunk 2]
```diff
@@ -11,22 +11,22 @@
...
+        if (!$container->hasDefinition(ManagerInterface::class)) {
             return;
         }
 
         $elementServices = $container->findTaggedServiceIds('admin.manager.visitor');
         foreach (array_keys($elementServices) as $id) {
             $visitor = $container->findDefinition($id);
-
-            $container->findDefinition('admin.manager')->addMethodCall('accept', [$visitor]);
+            $container->findDefinition(ManagerInterface::class)->addMethodCall('accept', [$visitor]);
```
Comment (szymach) [inline DependencyInjection/Compiler/ManagerVisitorPass.php:29]: Isn't this a duplication of the `accept` call in the Manager constructor?

File: DependencyInjection/Compiler/ResourceRepositoryPass.php [hunk 3]
```diff
@@ -4,21 +4,16 @@
...
 {
     public function process(ContainerBuilder $container): void
     {
-        if ($container->hasExtension('fsi_resource_repository')) {
-            $loader = new XmlFileLoader(
-                $container,
-                new FileLocator(__DIR__ . '/../../Resources/config')
-            );
-            $loader->load('context/resource.xml');
+        if (false === $container->hasExtension('fsi_resource_repository')) {
```
Comment (szymach) [inline DependencyInjection/Compiler/ResourceRepositoryPass.php:15]: I think it is better to parse service definition files only when the relevent extension exists, a bit less overhead.

File: Form/CollectionTypeExtension.php [hunk 4]
```diff
@@ -11,12 +11,12 @@
...
 class CollectionTypeExtension extends AbstractTypeExtension
 {
-    public static function getExtendedTypes()
+    public static function getExtendedTypes(): iterable
     {
         return [CollectionType::class];
     }
 
-    public function getExtendedType()
+    public function getExtendedType(): string
```
Comment (szymach) [inline Form/CollectionTypeExtension.php:19]: Isn't this redundant now?

File: behat.yml [hunk 5]
```diff
@@ -27,8 +27,10 @@ default: &default
             kernel:
                 env: test
                 debug: true
-                path: features/fixtures/project/app/AppKernel.php
+                path: features/fixtures/project/src/AppKernel.php
+                class: FSi\AppKernel
                 bootstrap: features/fixtures/project/app/bootstrap.php
+        Caciobanu\Behat\DeprecationExtension: ~
```
Comment (szymach) [inline behat.yml:33]: You could also add the extension for making screenshots on failure?

File: DependencyInjection/Compiler/ResourceRepositoryPass.php (hunk 3 above)
Comment (chives) [reply DependencyInjection/Compiler/ResourceRepositoryPass.php:15]: all services are registered with one prototype

**Review by szymach — APPROVED** [review]

**Comment by chives** [comment]
@fsi-open this is only the first step towards 4.0

**Review by szymach — APPROVED** [review]

**Comment by chives** [comment]
@fsi-open

**Comment by chives** [comment]
@fsi-open

File: Admin/CRUD/Context/Request/DeleteRequestHandler.php [hunk 6]
```diff
@@ -16,17 +16,20 @@
...
 use Symfony\Component\HttpFoundation\Response;
 use Symfony\Component\Routing\RouterInterface;
 
+use function get_class;
+
 class DeleteRequestHandler implements HandlerInterface
 {
     /**
-     * @var HandlerInterface
+     * @var BatchFormValidRequestHandler
```
Comment (szymach) [inline Admin/CRUD/Context/Request/DeleteRequestHandler.php:32]: Is this change neccessary? I'd avoid using specific classes if possible.

File: Behat/Context/AdminContext.php [hunk 7]
```diff
@@ -12,22 +12,35 @@
...
+use FriendsOfBehat\SymfonyExtension\Mink\MinkParameters;
 use FSi\Bundle\AdminBundle\Admin\AbstractElement;
 use FSi\Bundle\AdminBundle\Admin\Element;
 use FSi\Bundle\AdminBundle\Admin\ManagerInterface;
 use FSi\Bundle\AdminBundle\Behat\Page\AdminPanel;
 use FSi\Bundle\AdminBundle\Behat\Page\Page;
+use RuntimeException;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+use function var_dump;
```
Comment (szymach) [inline Behat/Context/AdminContext.php:26]: Leftover.

File: Behat/Context/FiltersContext.php [hunk 8]
```diff
@@ -12,28 +12,36 @@
...
 use FSi\Bundle\AdminBundle\Admin\CRUD\ListElement as AdminListElement;
 use FSi\Bundle\AdminBundle\Behat\Element\Filters;
 use FSi\Bundle\AdminBundle\Behat\Element\ListElement;
 use FSi\Bundle\AdminBundle\Behat\Element\ListResultsElement;
-use FSi\Bundle\AdminBundle\Behat\Page\DefaultPage;
+use FSi\Bundle\AdminBundle\Behat\Page\AdminPanel;
 use FSi\Component\DataSource\DataSourceInterface;
+use FSi\FixturesBundle\DataSource\DataSourceFactory;
+
+use function var_dump;
```
Comment (szymach) [inline Behat/Context/FiltersContext.php:26]: Leftover.

File: Behat/Page/CategoryList.php [hunk 9]
```diff
@@ -11,16 +11,24 @@
...
-    protected $path = '/admin/list/category';
-
-    protected function verifyPage(): void
+    public function verify(array $urlParameters = []): void
     {
-        if (false === $this->has('css', '#page-header:contains("List of elements")')) {
-            throw new UnexpectedPageException(sprintf("%s page is missing \"List of elements\" header", $this->path));
+        parent::verify($urlParameters);
+
+        var_dump($this->getDocument()->find('css', '#page-header')->getOuterHtml());
```
Comment (szymach) [inline Behat/Page/CategoryList.php:22]: Leftover.

File: Controller/ReorderController.php [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+
+use function get_class;
+use function sprintf;
+
+class ReorderController
```
Comment (szymach) [inline Controller/ReorderController.php:31]: `ReorderTreeController`?

File: Event/AdminEvent.php [hunk 11]
```diff
@@ -49,6 +38,10 @@ public function getRequest(): Request
         return $this->request;
     }
 
+    /**
+     * @return bool
+     * @phpstan-impure
```
Comment (szymach) [inline Event/AdminEvent.php:43]: How so?

File: composer.json [hunk 12]
```diff
@@ -19,61 +19,67 @@
...
-        "symfony/twig-bundle": "^3.4|^4.0",
-        "symfony/validator": "^3.4|^4.0",
-        "fsi/datagrid-bundle" : "^2.0",
-        "fsi/datasource-bundle": "^2.0",
-        "knplabs/knp-menu-bundle": "^2.0",
+        "doctrine/doctrine-bundle": "^2.0.3",
+        "doctrine/persistence": "^1.3|^2.0",
+        "doctrine/orm": "^2.8",
+        "sensio/framework-extra-bundle": "^5.5.7|^6.0",
+        "symfony/asset": "^4.4|^5.0",
```
Comment (szymach) [inline composer.json:28]: We have an LTS for `5.x`, so I'd require `4.4|5.4`

File: Event/AdminEvent.php (hunk 11 above)
Comment (chives) [reply Event/AdminEvent.php:43]: phpstan does not understand that two calls to `hasResponse` can return different values i.e. if there was `setResponse` call between them

File: Admin/CRUD/Context/Request/DeleteRequestHandler.php (hunk 6 above)
Comment (chives) [reply Admin/CRUD/Context/Request/DeleteRequestHandler.php:32]: `DeleteRequestHandler` strctly requires `BatchFormValidRequestHandler`

Comment (szymach) [reply Admin/CRUD/Context/Request/DeleteRequestHandler.php:32]: It may make it more difficult to replace it with an alternate implementation if need be.

**Comment by chives** [comment]
@fsi-open

File: composer.json [hunk 13]
```diff
@@ -24,28 +24,27 @@
...
+        "symfony/framework-bundle" : "^4.4|^5.4",
+        "symfony/http-foundation": "^4.4|^5.4",
+        "symfony/intl" : "^4.4|^5.4",
+        "symfony/options-resolver": "^4.4|^5.4",
+        "symfony/proxy-manager-bridge": "^4.4|^5.4",
+        "symfony/translation": "^4.4.12|^5.4",
+        "symfony/twig-bridge": "^4.4|^5.4",
+        "symfony/twig-bundle": "^4.4|^5.4",
+        "symfony/validator": "^4.4|^5.4",
         "twig/twig": "^2.13"
```
Comment (szymach) [inline composer.json:48]: Can we also allow `twig` 3.0?

File: composer.json [hunk 14]
```diff
@@ -70,12 +69,12 @@
         "sebastian/comparator": "^3.0",
         "sebastian/exporter": "^3.1",
         "squizlabs/php_codesniffer": "^3.5",
-        "symfony/debug": "^4.4|^5.0",
-        "symfony/dom-crawler": "^4.4.12|^5.0",
-        "symfony/monolog-bridge": "^4.4|^5.0",
+        "symfony/debug": "^4.4|^5.4",
+        "symfony/dom-crawler": "^4.4.12|^5.4",
+        "symfony/monolog-bridge": "^4.4|^5.4",
         "symfony/monolog-bundle": "^3.0",
```
Comment (szymach) [inline composer.json:75]: Might be a good idea to require the highest current version.

File: composer.json (hunk 13 above)
Comment (chives) [reply composer.json:48]: not yet

**Comment by chives** [comment]
@fsi-open

**Review by szymach — APPROVED** [review]

**Pull request merged**